### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by dverdin**
A lot of modification from RENATER's work on Universalistes overt the last eight months.

I commit all in one big commit which is hard to review but is at least explicit about all  changes.

To sum up:
        - Specialized roles such as postfix have been introduced, to ease the future usage of other tools.
        - Most of what is not totally Sympa-specific (such as web server virtual hosts configuration) have been moved to dedicated roles: postfix, apache, etc.
        - As we did it for our needs, a "monitoring" role has been added. It sends a mail to the seerver to check whether the mails work or not. It exposes a web service to check the state of the monitoring.
        - Patches for 6.2.32 have been added: password preservation, ldap alias manager and and auto-signoff fix. Auto-signoff fix will be pushed to the main Sympa project if it is not fixed already. Patch for LDAP alias management, is based on postfix virtual aliases.
        - Parts of the environment are not located in a specific "private" repository, to allow to version these in dedicated repostitories while keeping a common base for all environlents. For example, tasks could be common to several environments, but virtual hosts would differ.
        - Variables have been renamed to namespaces: 'sympa' for sympa, 'common' for the rest. Alows to add other namespaces ('wiki' and 'limesurvey' in the case of Universalistes), easing the work on a group of applications instead of Sympa alone.
        - Adding "hash_behaviour=merge" in ansible.cfg to allow to redefine only specific parameters in environments without having to repeat the whole hash in the "defalts" repository of a role.
        - Adding a "datasources" PHP script. I's a tool to check all datasources of a Sympa instance. Usefull to have a synthesis of where and how your server provisions its lists.
        - Sympa web configuration is now in the "conf-enabled" directory of Apache, making it automatically available to all vhosts.
        - Adding action_plugins/jin_ver.py to check jinja2 version at script startup.
        - Password and certificates encryption are enabled using keyczar.

### CODE REVIEW

**Comment by racke** [comment]
@dverdin I'm going to take a look at it, but it will take time based on the sheer amount of it.
I guess a start with picking up some quick wins first to cherry-pick.

Questions:

Why did you choose PHP scripts? Looks like a bunch of extra dependencies for just a small part of the whole project. 

What kind of monitoring solution you aim at?

**Comment by dverdin** [comment]
Hi,
Yes, I know it is hard to review. In the future, I'll make my best to publish sooner.

About PHP, these may be considered optional. It is only convenience scripts that Étienne wrote and he's better at PHP than at Perl. Maybe these would be better put out of the playbook, as contribution or something.

When a mail sent comes back to the server, a status is written in a file accessible through the web. That way, any external monitoring solution can check it. In our case, it is outsourced to a network operations center.

**Comment by dverdin** [comment]
Correction: Only the datasources analysis is in PHP. the monitoring scripts are in Perl. Sorry, I forgot that Étienne actually assumed that Sympa servers could be missing PHP dependencies and thus rewrote his monitoring in Perl.
I'll fix the PR description.

**Comment by dverdin** [comment]
Actually, it was already written... :-D

**Comment by racke** [comment]
I really don't think this monitoring stuff belongs here. Every existing monitoring can do checks on processes and such natively and the scripts don't have any room for flexibility (hardcoded paths).

**Comment by dverdin** [comment]
Yes, I agreed with that. I'll make it a contribution.
I'll remove it from the playbook. Same for the datasources analysis.

**Comment by dverdin** [comment]
But be advised that this check is for global messenging check. Not processes. We want to know whether the messages chain leading to our Sympa server works or not. As we know, it is not only a matter of local processes. Taht would be too easy... ;-)

**Comment by racke** [comment]
OK, I see - but all that comes with neglible documentation at all as far as I can see. Maybe I should go through your bullet points in that PR and see if I can cherry pick some neat stuff.

**Comment by dverdin** [comment]
Once again: I agree with removing the monitoring (and datasources analysis) from the playbook.

**Comment by racke** [comment]
OK I suggest that I'm going to test your changes next and report any apparent problems back - and move out the monitoring related parts (and maybe turn into a more meaningful check recognized by Nagios and friends).

**Comment by dverdin** [comment]
Cool !
The monitoring part is a sub-task in the postfix role.
Have a nice week-end!

File: README.md
Comment (Scriptkiddi) [inline README.md:164]: Could we not use a variable here and do the look up as a default for the variable

File: Vagrantfile
Comment (Scriptkiddi) [inline Vagrantfile:15]: why is there a hardcoded ip in here? I guess it makes some sense for Vagrant but is there a possibility of avoding it?

File: environments/template/group_vars/all.yml
Comment (Scriptkiddi) [inline environments/template/group_vars/all.yml]: those should be example emails

File: roles/apache/tasks/main.yml
Comment (Scriptkiddi) [inline roles/apache/tasks/main.yml:8]: could we move the apache stuff into an apache.yml and include it, would be nicer instead of having it in the main.yml so it can easier be extended with lets say nginx

File: roles/apache/templates/apache_vhost.conf.j2
Comment (Scriptkiddi) [inline roles/apache/templates/apache_vhost.conf.j2]: why is there a doku wiki and limewire entry in here?

File: roles/sympa/tasks/sympa_db_pg.yml
Comment (Scriptkiddi) [inline roles/sympa/tasks/sympa_db_pg.yml:29]: why no vault for postgres?

File: Vagrantfile
Comment (dverdin) [reply Vagrantfile:15]: It's only for usage with Vagrant indeed: we need to give the virtual server an Ip in order to know where it will answer. That' way, we know how to setup our /etc/hosts to be able to contact the server running locally.
I think the original author put it in the Vagrantfile because it is the only context where we setup IP from the project.
The thing is: before running Vagrant, we don't have an inventory. And we can't run the playbook without the inventory. So I really don't know how to use, e.g. the environment, to setup the IP.
Any suggestion welcome, though.

File: README.md
Comment (dverdin) [reply README.md:164]: Sorry, I do not understand what you want to do.
Because db.app_password is a variable.
Do you want to write something like that? 

```
db:
    type: mysql
    app_user: sympa
    app_password: sympa_app_password_content
```
and somewhere else:

```
sympa_app_password_content:  "{{ lookup('file',inventory_dir+'/private/password/sympa_db_app_password') }}"
```

Is it your idea?

File: environments/template/group_vars/all.yml
Comment (dverdin) [reply environments/template/group_vars/all.yml]: Oh yes. Forgot them.
(actually, "webmail.com" does not seem to be a thing but still).

File: roles/apache/tasks/main.yml
Comment (dverdin) [reply roles/apache/tasks/main.yml:8]: Something like this?

```
roles/
└── webserver/
    ├── apache.yml
    ├── main.yml
    └── nginx.yml
```

File: roles/apache/templates/apache_vhost.conf.j2
Comment (dverdin) [reply roles/apache/templates/apache_vhost.conf.j2]: Missing cleanup. sorry. It should not. I'll remove it.
But as an explanation: because in Universalistes (groupes.renater.fr) we use dokuwiki and limesurvey as Sympa subrogates.

File: roles/sympa/tasks/sympa_db_pg.yml
Comment (dverdin) [reply roles/sympa/tasks/sympa_db_pg.yml:29]: Oh, because Postgres was added by @racke _before_ I added the vault.
This will be fixed.

File: roles/apache/tasks/main.yml
Comment (Scriptkiddi) [reply roles/apache/tasks/main.yml:8]: yeah excactly

File: roles/apache/templates/apache_vhost.conf.j2
Comment (Scriptkiddi) [reply roles/apache/templates/apache_vhost.conf.j2]: i figured :D

File: README.md
Comment (Scriptkiddi) [reply README.md:164]: yeah perfect it allows to supply a password without putting it in a file

File: roles/sympa/tasks/sympa_db_pg.yml
Comment (dverdin) [reply roles/sympa/tasks/sympa_db_pg.yml:29]: Fixed by commit 7946e1f7a66a739d412a06d06f18b00f9ddff8ad.

File: roles/apache/tasks/main.yml
Comment (dverdin) [reply roles/apache/tasks/main.yml:8]: Would you agree to make this change (which is a bit deep) in another PR? I'd like to reorganize on a stable basis.

File: environments/template/group_vars/all.yml
Comment (dverdin) [reply environments/template/group_vars/all.yml]: This was fixed by commit ecc43e6fe1d2b98743bb952c8c7f9343159b2c62.

File: README.md
Comment (dverdin) [reply README.md:164]: As for the webserver issue, I'd like to introduce it in another PR.? Because, in order to make it correctly, We should also modify the script generating the passwords, encrypted by the vault.
I can create a distinct issue if you like.

File: roles/apache/templates/apache_vhost.conf.j2
Comment (dverdin) [reply roles/apache/templates/apache_vhost.conf.j2]: This was fixed in commit 8ba22d0.
I was suspecting it for a while but I'm pretty sure now: in order to have playbooks working in a multi-tools environment, we need to get rid of templates in most cases and edit parts of the config files from tasks, using lineinfile and blockinfile.

File: roles/apache/tasks/main.yml
Comment (Scriptkiddi) [reply roles/apache/tasks/main.yml:8]: sure I currently try to find time to add something that works with exim and nginx to this role, so If i get around to it I will do that change anyway :D

File: roles/apache/templates/apache_vhost.conf.j2
Comment (racke) [reply roles/apache/templates/apache_vhost.conf.j2]: > This was fixed in commit [8ba22d0](https://github.com/sympa-community/sympa-ansible/commit/8ba22d093ffe124c84d0182c439ebd620500183e).
> I was suspecting it for a while but I'm pretty sure now: in order to have playbooks working in a multi-tools environment, we need to get rid of templates in most cases and edit parts of the config files from tasks, using lineinfile and blockinfile.

Not necessarily - many servers provide include statements in their configuration set which can solve this in some cases.

Comment (dverdin) [reply roles/apache/templates/apache_vhost.conf.j2]: It would be very nice !

**Comment by dverdin** [comment]
Dear all,
I know I made you wait for a long time for this PR but did you have the time to have an idea about its pertinence? I have a few more improvements - way more atomistic - I would like to push and, as they are based on this, I can't do it right now.
So, anyone against the merge? :-D

EDIT: Typo.

**Comment by racke** [comment]
:-1: I don't think that is a good idea to merge it at that point. It would be good to get a few low hanging fruit out of it - e.g. what about the patches you are talking about. Also the diff for README.md is very small compared to the amount of changes.

**Comment by dverdin** [comment]
> -1 I don't think that is a good idea to merge it at that point. It would be good to get a few low hanging fruit out of it - e.g. what about the patches you are talking about. Also the diff for README.md is very small compared to the amount of changes.

Ok, If you feel this way, I'll wait.
I can work on the readme.MD. Documenting the available parameters for example.
When you speak about patches, do you mean: the one I applied to Sympa 6.2.32 during install? I should have proposed them to the main sympa project. Didn't get the time yet.

**Comment by dverdin** [comment]
Dear all,
It's been a month now. We won't go any further if we don't merge; could we please merge and move on?
This play book is  what we use to deploy in production (and the three next PR improve it) I'm fairly confident it works and is better organized than the current master.
Cheers,
David

**Comment by racke** [comment]
Honestly, I'm not really in favour to merging this refactoring ... especially before you leave for two weeks and not being able to fxi possible problems caused by it. 

Also I don't see the README modifications which surely should accompany such a big update.

In addition I would like to see a breakdown of the Sympa role - that should only install and configure Sympa and *not* databases, webservers and what-not.

I'm happy if we can discuss that further when you are back.

**Comment by dverdin** [comment]
Yes, we can discuss when I'm back.
But I'm not giving up months of work because I don't submit them the way they should.
I also want to add more independence between roles but what I did in this PR is a prerequisite for going further.
See you in two weeks!

**Comment by racke** [comment]
Also you didn't update the PR with the new download location as per 5e28e737211a733bb44c0d4132c14b6bb79aa03b.

**Comment by racke** [comment]
Error: 

TASK [apache : Put Apache vhost certs] ***************************************************************************************************************************
fatal: [local-sympa]: FAILED! => {"msg": "The conditional check 'item.ssl.enabled' failed. The error was: error while evaluating conditional (item.ssl.enabled): 'dict object' has no attribute 'ssl'\n\nThe error appears to have been in '/home/racke/sympa-community/renater/roles/apache/tasks/main.yml': line 30, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Put Apache vhost certs\n  ^ here\n"}

**Comment by racke** [comment]
Sorry, but I'm not going to try to analyze this _set_fact_ task which produces the _apache_vhosts_ variable. @dverdin Please fix this or provide more information.

**Review by racke — CHANGES_REQUESTED** [review]
See my various concerns, there is not a chance that I approve this pull request in its current state.

**Comment by dverdin** [comment]
> Error:
> 
> TASK [apache : Put Apache vhost certs] ***************************************************************************************************************************
> fatal: [local-sympa]: FAILED! => {"msg": "The conditional check 'item.ssl.enabled' failed. The error was: error while evaluating conditional (item.ssl.enabled): 'dict object' has no attribute 'ssl'\n\nThe error appears to have been in '/home/racke/sympa-community/renater/roles/apache/tasks/main.yml': line 30, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Put Apache vhost certs\n ^ here\n"}

Hi racke,
Can I see the group_vars/all.yml file from your environment please?

**Comment by dverdin** [comment]
> Also you didn't update the PR with the new download location as per [5e28e73](https://github.com/sympa-community/sympa-ansible/commit/5e28e737211a733bb44c0d4132c14b6bb79aa03b).

Done!

**Comment by racke** [comment]
> > Error:
> > TASK [apache : Put Apache vhost certs] ***************************************************************************************************************************
> > fatal: [local-sympa]: FAILED! => {"msg": "The conditional check 'item.ssl.enabled' failed. The error was: error while evaluating conditional (item.ssl.enabled): 'dict object' has no attribute 'ssl'\n\nThe error appears to have been in '/home/racke/sympa-community/renater/roles/apache/tasks/main.yml': line 30, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Put Apache vhost certs\n ^ here\n"}
> 
> Hi racke,
> Can I see the group_vars/all.yml file from your environment please?

OK, looks like it was incomplete.

**Comment by dverdin** [comment]
This is what I reached tonight in the README file.
It still misses informations about the development and maybe detailed informations about the parameters (for example, how to switch from Pg to mysql)
But it might clarify things already.
I'll start again tomorrow: https://github.com/dverdin/sympa-ansible/blob/master/README.md

**Comment by racke** [comment]
Good.

> Ansible-tools promotes a setup for encrypting secrets that is different from the Ansible Vault. The Ansible vault feature encrypts an entire .yml file with variable names and values, whereas the ansible-tools approach encrypts just the values en decrypts them just before they are needed. Both use the python-keyczar for encryption.

This is not correct. You can encrypt single values in a .yml file with the standard Ansible vault. Also it doesn't use python-keyczar.

**Comment by dverdin** [comment]
Indeed. I spent some time experimenting with ansible-vault today and it looks like it can indeed handle single variables now.
That, added to the ability to store the vault password in a file, seems to fill all my needs.
We'll be able to switch from keyczar to ansible vault easily after the merge, I think. I already started working on the environment generation script.
I just need to make sure how to use encrypted files in the playbooks.

**Comment by racke** [comment]
From my experience, Ansible can figure out by itself when YAML files are completely or partially encrypted by Ansible vault. So actually playbooks should work as is.

**Comment by dverdin** [comment]
OK, here is the updated README: https://github.com/sympa-community/sympa-ansible/blob/5ce712b611ba87f077ebc9953746ff7d963c5fc9/README.md

Any other remarks before the merge?

**Comment by racke** [comment]
I did a lot of remarks here already ... going to check them after the merge. I expect that you deal with them ASAP. Especially the Keyczar problem.

**Comment by dverdin** [comment]
And I certainly appreciate what you did here. At least, I can't sink into self-satisfaction... ;-)
I have a pretty good idea about how to change towards the Ansible vault.

**Pull request merged**
